Katello/katello
# Fixes #37455 - Create Katello push repositories as needed at container push time, handle non-auth validation.

#### What are the changes introduced in this pull request?

- When using podman to push containers to Katello, ensure that the container name is of a valid format:
  - <org_label>/<product_label>/<container_name>[:tag] (down-case all letters)
  -  id/<org_id>/<product_id>/<container_name>[:tag]
- Limits to container names:
  - Org and product fields must always correspond to a valid, existing org and product, respectively. No fields can be blank or omitted.
  - Org and product labels must be downcased. Container names must be lowercase and contain only letters and underscores (podman rejects these automatically).
  - Org and product labels cannot be ambiguous (for example 'foo_bar' and 'Foo_Bar' both map to 'foo_bar'). For ambiguous orgs or products, the user is directed to use the id format.
  - Org and product IDs must be integers with no leading zeros.
  - A user must stick to a specific format for each org, product, name combination. Pushes are not permitted to alternate between label and id or vice versa.
  - If a label format push becomes ambiguous due to external changes, the user will be directed to destroy either the offending org/product or will be directed to destroy the existing push container and re-push using the id format.
- Create a new push repository matching the container name under the appropriate product where required.
- Update existing push repos where required.
- Run content indexing after completion of the container upload.
- Ensure the root repository fields are correct after upload:
  - name: matches container_name
  - label: matches container_name
  - content_type: docker
  - is_container_push (new field): true
  - container_push_name (new field): name used in podman command e.g. "id/0/0/foo"
  - container_push_name_format (new field): name format used, `label` or `id` 
- Ensure that the instance repository fields are correct after upload:
  - relative_path: matches the container_push_name above
  - container_repository_name: matches the container_push_name above
  - environment_id: correct for product
  - content_view_version_id: correct for product
  - latest_version_href: pulp latest version href
  - pulp_href: pulp href

#### Considerations taken when implementing this change?

#### What are the testing steps for this pull request?

1. Pull a few different podman images:
 ```
podman pull quay.io/aptible/hello-world
 ```
 2. Launch katello and create an empty product. Log in with podman:
 ```
 podman login <url>
 ```
 3. Get the image ID for one of the test podman images. Push that image to katello using the correct organization label and product label. Make sure to down-case the labels or podman will reject the container name:
 ```
 podman image list
 podman push <img_id> <url>/<organization_label>/<product_label>/<container_name>
 ```
4. Use the rake console to ensure the root repository and repository were created properly:
```
bundle exec rake console
::Katello::Products.find(<id>).root_repositories.where(label: "<container_name>").first
::Katello::Products.find(<id>).root_repositories.where(label: "<container_name>").first.library_instance
```
5. Repeat steps 3 and 4 using the id push format. You will need to change the container name.
```
podman push <img_id> <url>/id/<organization_id>/<product_id>/<container_name>
```
6. Repeat steps 3 and 4 with a [:tag] field. This should work as usual.
7. Repeat steps 3 and 4 while trying every error in the above 'Limits to container names' list. This is an extensive but not exhaustive list of all the error handling included. List duplicated below:
  - Org and product fields must always correspond to a valid, existing org and product, respectively. No fields can be blank or omitted.
  - Org and product labels must be downcased. Container names must be lowercase and contain only letters and underscores (podman rejects these automatically).
  - Org and product labels cannot be ambiguous (for example 'foo_bar' and 'Foo_Bar' both map to 'foo_bar'). For ambiguous orgs or products, the user is directed to use the id format.
  - Org and product IDs must be integers with no leading zeros.
  - A user must stick to a specific format for each org, product, name combination. Pushes are not permitted to alternate between label and id or vice versa for a given container.
  - If a label format push becomes ambiguous due to external changes, the user will be directed to destroy either the offending org/product or will be directed to destroy the existing push container and re-push using the id format.
8. Validate the test cases for coverage and accuracy.
